Exception occurred while broadcasting an empty result

### Describe the bug

An exception occurs when broadcasting an empty result. I think this is due to the extra bytes being written in the encode compressed stream, causing it to be determined not empty in `CometExec$.decodeBatches`.

error detail:

```
Caused by: java.io.IOException: Unexpected end of input. Missing schema.
	at org.apache.arrow.vector.ipc.ArrowStreamReader.readSchema(ArrowStreamReader.java:205)
	at org.apache.arrow.vector.ipc.ArrowReader.initialize(ArrowReader.java:185)
	at org.apache.arrow.vector.ipc.ArrowReader.ensureInitialized(ArrowReader.java:176)
	at org.apache.arrow.vector.ipc.ArrowReader.getVectorSchemaRoot(ArrowReader.java:73)
	at org.apache.comet.vector.StreamReader.<init>(StreamReader.scala:39)
	at org.apache.spark.sql.comet.execution.shuffle.ArrowReaderIterator.<init>(ArrowReaderIterator.scala:36)
	at org.apache.spark.sql.comet.CometExec$.decodeBatches(operators.scala:135)
	at org.apache.spark.sql.comet.CometBatchRDD.$anonfun$compute$1(CometBroadcastExchangeExec.scala:249)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.mutable.ArrayOps$ofRef.flatMap(ArrayOps.scala:198)
	at org.apache.spark.sql.comet.CometBatchRDD.compute(CometBroadcastExchangeExec.scala:249)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
```

### Steps to reproduce

broadcast `tbl_a` in CometBroadcastExchangeExec test case

```
SELECT /*+ BROADCAST(tbl_a) */ tbl_a._1, tbl_b._2 FROM tbl_a JOIN tbl_b " +
                "WHERE tbl_a._1 > tbl_a._2 LIMIT 2
```

### Expected behavior

_No response_

### Additional context

_No response_

--- a/spark/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/CometShuffleManager.scala
+++ b/spark/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/CometShuffleManager.scala
@@ -52,7 +52,7 @@ class CometShuffleManager(conf: SparkConf) extends ShuffleManager with Logging {
         " Shuffle will continue to spill to disk when necessary.")
   }
 
-  private val sortShuffleManager = new SortShuffleManager(conf);
+  private val sortShuffleManager = new SortShuffleManager(conf)
 
   /**
    * A mapping from shuffle ids to the task ids of mappers producing output for those shuffles.

--- a/spark/src/main/scala/org/apache/spark/sql/comet/operators.scala
+++ b/spark/src/main/scala/org/apache/spark/sql/comet/operators.scala
@@ -81,7 +81,11 @@ abstract class CometExec extends CometPlan {
 
       out.flush()
       out.close()
-      Iterator((count, cbbos.toChunkedByteBuffer))
+      if (out.size() > 0) {
+        Iterator((count, cbbos.toChunkedByteBuffer))
+      } else {
+        Iterator((count, new ChunkedByteBuffer(Array.empty[ByteBuffer])))
+      }
     }
   }
 

--- a/spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
+++ b/spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
@@ -59,22 +59,28 @@ class CometExecSuite extends CometTestBase {
     withSQLConf(CometConf.COMET_EXEC_BROADCAST_ENABLED.key -> "true") {
       withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_a") {
         withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_b") {
-          val df = sql(
-            "SELECT tbl_a._1, tbl_b._2 FROM tbl_a JOIN tbl_b " +
-              "WHERE tbl_a._1 > tbl_a._2 LIMIT 2")
+          Seq(true, false).foreach { emptyBroadcast =>
+            val broadcastTable = if (emptyBroadcast) "tbl_a" else "tbl_b"
+            val df = sql(
+              s"SELECT /*+ BROADCAST($broadcastTable) */ tbl_a._1, tbl_b._2 FROM tbl_a JOIN tbl_b " +
+                "WHERE tbl_a._1 > tbl_a._2 LIMIT 2")
 
-          val nativeBroadcast = find(df.queryExecution.executedPlan) {
-            case _: CometBroadcastExchangeExec => true
-            case _ => false
-          }.get.asInstanceOf[CometBroadcastExchangeExec]
+            val nativeBroadcast = find(df.queryExecution.executedPlan) {
+              case _: CometBroadcastExchangeExec => true
+              case _ => false
+            }.get.asInstanceOf[CometBroadcastExchangeExec]
 
-          val numParts = nativeBroadcast.executeColumnar().getNumPartitions
+            val numParts = nativeBroadcast.executeColumnar().getNumPartitions
 
-          val rows = nativeBroadcast.executeCollect().toSeq.sortBy(row => row.getInt(0))
-          val rowContents = rows.map(row => row.getInt(0))
-          val expected = (0 until numParts).flatMap(_ => (0 until 5).map(i => i + 1)).sorted
-
-          assert(rowContents === expected)
+            val rows = nativeBroadcast.executeCollect().toSeq.sortBy(row => row.getInt(0))
+            val rowContents = rows.map(row => row.getInt(0))
+            val expected = if (emptyBroadcast) {
+              Seq.empty
+            } else {
+              (0 until numParts).flatMap(_ => (0 until 5).map(i => i + 1)).sorted
+            }
+            assert(rowContents === expected)
+          }
         }
       }
     }

--- a/spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
+++ b/spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
@@ -59,28 +59,41 @@ class CometExecSuite extends CometTestBase {
     withSQLConf(CometConf.COMET_EXEC_BROADCAST_ENABLED.key -> "true") {
       withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_a") {
         withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_b") {
-          Seq(true, false).foreach { emptyBroadcast =>
-            val broadcastTable = if (emptyBroadcast) "tbl_a" else "tbl_b"
-            val df = sql(
-              s"SELECT /*+ BROADCAST($broadcastTable) */ tbl_a._1, tbl_b._2 FROM tbl_a JOIN tbl_b " +
-                "WHERE tbl_a._1 > tbl_a._2 LIMIT 2")
+          val df = sql(
+            "SELECT tbl_a._1, tbl_b._2 FROM tbl_a JOIN tbl_b " +
+              "WHERE tbl_a._1 > tbl_a._2 LIMIT 2")
 
-            val nativeBroadcast = find(df.queryExecution.executedPlan) {
-              case _: CometBroadcastExchangeExec => true
-              case _ => false
-            }.get.asInstanceOf[CometBroadcastExchangeExec]
+          val nativeBroadcast = find(df.queryExecution.executedPlan) {
+            case _: CometBroadcastExchangeExec => true
+            case _ => false
+          }.get.asInstanceOf[CometBroadcastExchangeExec]
 
-            val numParts = nativeBroadcast.executeColumnar().getNumPartitions
+          val numParts = nativeBroadcast.executeColumnar().getNumPartitions
 
-            val rows = nativeBroadcast.executeCollect().toSeq.sortBy(row => row.getInt(0))
-            val rowContents = rows.map(row => row.getInt(0))
-            val expected = if (emptyBroadcast) {
-              Seq.empty
-            } else {
-              (0 until numParts).flatMap(_ => (0 until 5).map(i => i + 1)).sorted
-            }
-            assert(rowContents === expected)
-          }
+          val rows = nativeBroadcast.executeCollect().toSeq.sortBy(row => row.getInt(0))
+          val rowContents = rows.map(row => row.getInt(0))
+          val expected = (0 until numParts).flatMap(_ => (0 until 5).map(i => i + 1)).sorted
+
+          assert(rowContents === expected)
+        }
+      }
+    }
+  }
+
+  test("CometBroadcastExchangeExec: empty broadcast") {
+    withSQLConf(CometConf.COMET_EXEC_BROADCAST_ENABLED.key -> "true") {
+      withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_a") {
+        withParquetTable((0 until 5).map(i => (i, i + 1)), "tbl_b") {
+          val df = sql(
+            "SELECT /*+ BROADCAST(a) */ *" +
+              " FROM (SELECT * FROM tbl_a WHERE _1 < 0) a JOIN tbl_b b" +
+              " ON a._1 = b._1")
+          val nativeBroadcast = find(df.queryExecution.executedPlan) {
+            case _: CometBroadcastExchangeExec => true
+            case _ => false
+          }.get.asInstanceOf[CometBroadcastExchangeExec]
+          val rows = nativeBroadcast.executeCollect()
+          assert(rows.isEmpty)
         }
       }
     }

